### ISSUE TITLE
Provide a way to measure graphic rendering time

### ISSUE THREAD

**Issue by squinkylabs** [VCVRack/Rack#1390]

**Comment by stoermelder** [comment]
I'm no way a graphics expert but I think this could be quite difficult. The engine thread executes each `process` synchronously what makes measurement easy. The application thread collects all drawing commands and sends the whole bunch to the graphics adapter which renders one frame asynchronously.

**Comment by AndrewBelt** [comment]
I can certainly measure the amount of time spent in `ModuleWidget::step()` and `draw()` functions, but there's a third, possibly more important function that is affected by draw "weight", which is `nvgEndFrame(vg)`, which cannot be reliably measured. On the other hand, since FramebufferWidget drawing happens in `step()`, it could be useful for monitoring FramebufferWidget time.

**Comment by squinkylabs** [comment]
I don't know how valuable it is, but I was thinking of just time spent in my step and draw functions. Because I can control that. True if you have 1000 FramebufferWidgets the time will be zero but nanogl/hardware may not like that. I guess I don't know much about how to performance breaks down between these different stages, and hence what is useful to measure. Probably should make some measurements of my own to learn more. Yeah, I agree we don't really need more user confusion. Ideally it would be "nice" if rack could suggest a lower frame rate, as many users seem the think that helps often, but yeah, no more things that people can compare for no good reason.

**Comment by squinkylabs** [comment]
I hope to write up some results soon. Measured the time I take just in Module::draw(). As with CPU usage, I set a target of 1% available time as a "quota" for a module. Most of mine are quite under that, but a couple were terrible, using more like 7% of available frame time. These turned out to be all modules that had my own widgets, and the widgets were drawing SVG without using a frame buffer. Pretty dumb, but fixing that made the bad ones at least 10X better. And you can even seen this when zooming or scrolling the browser.

**Comment by JerrySievert** [comment]
@squinkylabs do you have a patch against rack that enables those modifications?  I'd hate to do all of the work again if you have something already built.

**Comment by squinkylabs** [comment]
No, I have a simple timer that I put into my plugins source code. And at the moment it only measures time in draw(). https://github.com/squinkylabs/SquinkyVCV/blob/master/sqsrc/util/DrawTimer.h
